File: src/SilksongLauncher.Launcher/app/src/main/kotlin/dev/silksong/launcher/skins/core/SkinResourceCore.kt
package dev.silksong.launcher.skins.core

/** Pure, bounded admission/reference accounting. No lifetime or apply authority is granted here. */
class SkinResourceCore {
    companion object {
        const val TEXTURE_BYTES = 32L * 1024 * 1024
        const val PLAN_BYTES = 96L * 1024 * 1024
        const val PROCESS_BYTES = 224L * 1024 * 1024
        const val PLAN_ALLOCATIONS = 205
        const val PROCESS_ALLOCATIONS = 410
        // Bounds on bookkeeping, not texture budgets. Lifetime scheduling belongs to the next slice.
        const val MAX_PLANS = 3
        const val MAX_REFERENCES = 16384
        const val MAX_SOURCE_FACTS = 16384
    }
    private data class Content(val length: Long, val hash: String, val width: Long, val height: Long, val format: String)
    private fun content(v: VerifiedResourceSource) = Content(v.encodedLength, v.sha256, v.width, v.height, v.format)
    private fun text(v: String, max: Int) = v.isNotBlank() && v.length <= max
    private fun bytes(width: Long, height: Long): Long {
        require(width > 0 && height > 0)
        return Math.multiplyExact(Math.multiplyExact(width, height), 4L)
    }
    private fun valid(v: VerifiedResourceSource): Boolean =
        text(v.proofScope, 128) && text(v.canonicalSource, 4096) && v.encodedLength > 0 &&
            v.sha256.length == 64 && v.sha256.all { it in '0'..'9' || it in 'a'..'f' } &&
            v.format == "RGBA32" && v.usage == ResourceUsage.GAME && bytes(v.width, v.height) <= TEXTURE_BYTES

    /** Null means inconsistent, malformed, overflowing, or over-budget input. Never trusts cached totals. */
    fun totals(state: ResourceState): ResourceTotals? = try {
        require(state.nextId > 0 && state.allocations.size <= PROCESS_ALLOCATIONS * 2 && state.references.size <= MAX_REFERENCES)
        val ids = HashSet<Long>()
        fun id(v: Long) { require(v > 0 && v < state.nextId && ids.add(v)) }
        val byId = HashMap<Long, ResourceAllocation>()
        val plans = HashSet<String>()
        val canonical = HashSet<Pair<String, String>>()
        val contents = HashSet<Content>()
        val borrowed = HashSet<String>()
        val planBytes = HashMap<String, Long>()
        val planCounts = HashMap<String, Int>()
        var resident = 0L
        var owned = 0
        var borrowedCount = 0
        var facts = 0
        for (a in state.allocations) {
            id(a.id); byId[a.id] = a
            require(a.planIds.isNotEmpty() && a.planIds.size <= MAX_PLANS && a.planIds.toSet().size == a.planIds.size)
            for (p in a.planIds) { require(text(p, 128)); plans.add(p) }
            val size = bytes(a.width, a.height)
            when (a.ownership) {
                ResourceOwnership.OWNED -> {
                    require(a.borrowedKey == null && a.sources.isNotEmpty() && size <= TEXTURE_BYTES)
                    facts = Math.addExact(facts, a.sources.size); require(facts <= MAX_SOURCE_FACTS)
                    val c = content(a.sources.first()); require(contents.add(c))
                    for (v in a.sources) {
                        require(valid(v) && content(v) == c && v.width == a.width && v.height == a.height)
                        require(canonical.add(v.proofScope to v.canonicalSource))
                    }
                    resident = Math.addExact(resident, size); owned = Math.addExact(owned, 1)
                    for (p in a.planIds) {
                        planBytes[p] = Math.addExact(planBytes[p] ?: 0, size)
                        planCounts[p] = Math.addExact(planCounts[p] ?: 0, 1)
                    }
                }
                ResourceOwnership.BORROWED -> {
                    require(a.sources.isEmpty() && a.borrowedKey != null && text(a.borrowedKey, 4096) && borrowed.add(a.borrowedKey))
                    require(a.phase == ResourceAllocationPhase.READY)
                    borrowedCount = Math.addExact(borrowedCount, 1)
                }
            }
        }
        require(plans.size <= MAX_PLANS && owned <= PROCESS_ALLOCATIONS && borrowedCount <= PROCESS_ALLOCATIONS)
        require(planBytes.values.all { it <= PLAN_BYTES } && planCounts.values.all { it <= PLAN_ALLOCATIONS })
        for (r in state.references) {
            id(r.id)
            val a = byId[r.allocationId]
            require(a != null && r.planId in a.planIds)
        }
        val scratch = state.scratch
        if (scratch != null) {
            id(scratch.operation.operationId)
            val a = byId[scratch.operation.allocationId]
            require(a != null && a.ownership == ResourceOwnership.OWNED && scratch.operation.planId in a.planIds)
            require(scratch.bytes == bytes(a.width, a.height) && scratch.bytes <= TEXTURE_BYTES)
            require(a.phase == if (scratch.decodeCompleted) ResourceAllocationPhase.READY else ResourceAllocationPhase.RESERVED)
        }
        require(state.allocations.filter { it.phase == ResourceAllocationPhase.RESERVED }.all {
            scratch != null && !scratch.decodeCompleted && scratch.operation.allocationId == it.id
        })
        val peak = Math.addExact(resident, scratch?.bytes ?: 0)
        require(peak <= PROCESS_BYTES)
        ResourceTotals(resident, peak, owned, planBytes, planCounts)
    } catch (_: IllegalArgumentException) { null } catch (_: ArithmeticException) { null }

    fun decide(state: ResourceState, event: ResourceEvent): ResourceDecision {
        fun reject(code: String) = ResourceDecision(state, false, code)
        if (totals(state) == null) return reject("INVALID_STATE")
        return try {
            when (event) {
                is ResourceEvent.Acquire -> {
                    if (!text(event.planId, 128) || !valid(event.source)) return reject("INVALID_SOURCE")
                    val c = content(event.source)
                    val owned = state.allocations.filter { it.ownership == ResourceOwnership.OWNED }
                    if (owned.any { a -> a.sources.any { it.proofScope == event.source.proofScope &&
                            it.canonicalSource == event.source.canonicalSource && content(it) != c } }) return reject("CONTRADICTORY_SOURCE")
                    val existing = owned.firstOrNull { content(it.sources.first()) == c }
                    if (existing == null && state.scratch != null) return reject("SCRATCH_BUSY")
                    val allocationId = existing?.id ?: state.nextId
                    val referenceId = if (existing == null) Math.addExact(state.nextId, 1) else state.nextId
                    val reference = ResourceReference(referenceId, event.planId, allocationId, event.kind)
                    val commands = mutableListOf<ResourceCommand>(ResourceCommand.ReferenceGranted(reference))
                    val updated = ResourceAllocation(allocationId, ResourceOwnership.OWNED, event.source.width, event.source.height,
                        ((existing?.sources ?: emptyList()) + event.source).distinct(),
                        ((existing?.planIds ?: emptyList()) + event.planId).distinct(), phase = existing?.phase ?: ResourceAllocationPhase.RESERVED)
                    var scratch = state.scratch
                    var next = Math.addExact(referenceId, 1)
                    if (existing == null) {
                        val op = ResourceDecodeOperation(next, allocationId, event.planId)
                        next = Math.addExact(next, 1)
                        scratch = ResourceScratch(op, bytes(updated.width, updated.height))
                        commands.add(ResourceCommand.Decode(op, event.source))
                    }
                    admit(state, ResourceState(state.allocations.filterNot { it.id == allocationId } + updated,
                        state.references + reference, scratch, next), commands)
                }
                is ResourceEvent.AcquireBorrowed -> {
                    if (!text(event.planId, 128) || !text(event.borrowedKey, 4096)) return reject("INVALID_BORROWED")
                    bytes(event.width, event.height)
                    val existing = state.allocations.firstOrNull { it.ownership == ResourceOwnership.BORROWED && it.borrowedKey == event.borrowedKey }
                    if (existing != null && (existing.width != event.width || existing.height != event.height)) return reject("CONTRADICTORY_BORROWED")
                    val aid = existing?.id ?: state.nextId
                    val rid = if (existing == null) Math.addExact(state.nextId, 1) else state.nextId
                    val reference = ResourceReference(rid, event.planId, aid, event.kind)
                    val allocation = ResourceAllocation(aid, ResourceOwnership.BORROWED, event.width, event.height,
                        planIds = ((existing?.planIds ?: emptyList()) + event.planId).distinct(), borrowedKey = event.borrowedKey, phase = ResourceAllocationPhase.READY)
                    admit(state, ResourceState(state.allocations.filterNot { it.id == aid } + allocation, state.references + reference,
                        state.scratch, Math.addExact(rid, 1)), listOf(ResourceCommand.ReferenceGranted(reference)))
                }
                is ResourceEvent.ReleaseReference -> {
                    if (event.reference !in state.references) return reject("STALE_REFERENCE")
                    admit(state, ResourceState(state.allocations, state.references.filterNot { it == event.reference }, state.scratch, state.nextId))
                }
                is ResourceEvent.DecodeCompleted -> {
                    val scratch = state.scratch
                    if (scratch == null || scratch.operation != event.operation || scratch.decodeCompleted) return reject("STALE_DECODE")
                    val allocations = state.allocations.map { a -> if (a.id != event.operation.allocationId) a else
                        ResourceAllocation(a.id, a.ownership, a.width, a.height, a.sources, a.planIds, a.borrowedKey, ResourceAllocationPhase.READY) }
                    admit(state, ResourceState(allocations, state.references, scratch.copy(decodeCompleted = true), state.nextId),
                        listOf(ResourceCommand.ReleaseUploadScratch(event.operation)))
                }
                is ResourceEvent.ScratchReleased -> {
                    val scratch = state.scratch
                    if (scratch == null || scratch.operation != event.operation || !scratch.decodeCompleted) return reject("STALE_SCRATCH")
                    admit(state, ResourceState(state.allocations, state.references, null, state.nextId))
                }
            }
        } catch (_: IllegalArgumentException) { reject("INVALID_EVENT") } catch (_: ArithmeticException) { reject("OVERFLOW") }
    }
    private fun admit(prior: ResourceState, next: ResourceState, commands: List<ResourceCommand> = emptyList()): ResourceDecision =
        if (totals(next) == null) ResourceDecision(prior, false, "LIMIT_OR_INCONSISTENT")
        else ResourceDecision(next, true, "ACCEPTED", commands)
}

File: tools/hollow-knight-patches/src/skins/core/SkinResourceCore.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace DualSouls.Skins.HollowKnight.Core
{
    // Pure bounded admission/reference accounting. Does not grant lifetime or apply authority.
    public sealed class SkinResourceCore
    {
        public const long TEXTURE_BYTES = 32L * 1024 * 1024;
        public const long PLAN_BYTES = 96L * 1024 * 1024;
        public const long PROCESS_BYTES = 224L * 1024 * 1024;
        public const int PLAN_ALLOCATIONS = 205;
        public const int PROCESS_ALLOCATIONS = 410;
        // Bookkeeping bounds, not texture budgets. Lifetime scheduling is a separate slice.
        public const int MAX_PLANS = 3;
        public const int MAX_REFERENCES = 16384;
        public const int MAX_SOURCE_FACTS = 16384;
        private sealed record Content(long Length, string Hash, long Width, long Height, string Format);
        private static Content Identity(VerifiedResourceSource v) => new(v.EncodedLength, v.Sha256, v.Width, v.Height, v.Format);
        private static bool Text(string v, int max) => !string.IsNullOrWhiteSpace(v) && v.Length <= max;
        private static void Require(bool condition) { if (!condition) throw new ArgumentException("Invalid resource value"); }
        private static long Bytes(long width, long height) { Require(width > 0 && height > 0); return checked(checked(width * height) * 4L); }
        private static bool Valid(VerifiedResourceSource v) => v != null && Text(v.ProofScope, 128) && Text(v.CanonicalSource, 4096) &&
            v.EncodedLength > 0 && v.Sha256 != null && v.Sha256.Length == 64 && v.Sha256.All(c => c >= '0' && c <= '9' || c >= 'a' && c <= 'f') &&
            v.Format == "RGBA32" && v.Usage == ResourceUsage.GAME && Bytes(v.Width, v.Height) <= TEXTURE_BYTES;
        private static bool Kind(ResourceReferenceKind k) => k == ResourceReferenceKind.TARGET || k == ResourceReferenceKind.MATERIAL || k == ResourceReferenceKind.ROLLBACK;

        // Null means inconsistent, malformed, overflowing or over-budget input. No cached totals are trusted.
        public ResourceTotals Totals(ResourceState state)
        {
            try
            {
                Require(state != null && state.NextId > 0 && state.Allocations.Count <= PROCESS_ALLOCATIONS * 2 && state.References.Count <= MAX_REFERENCES);
                var ids = new HashSet<long>();
                void Id(long value) => Require(value > 0 && value < state.NextId && ids.Add(value));
                var byId = new Dictionary<long, ResourceAllocation>();
                var plans = new HashSet<string>(); var canonical = new HashSet<(string, string)>();
                var contents = new HashSet<Content>(); var borrowed = new HashSet<string>();
                var planBytes = new Dictionary<string, long>(); var planCounts = new Dictionary<string, int>();
                long resident = 0; int owned = 0, borrowedCount = 0, facts = 0;
                foreach (var a in state.Allocations)
                {
                    Require(a != null); Id(a.Id); byId[a.Id] = a;
                    Require(a.Phase == ResourceAllocationPhase.RESERVED || a.Phase == ResourceAllocationPhase.READY);
                    Require(a.PlanIds.Count > 0 && a.PlanIds.Count <= MAX_PLANS && a.PlanIds.Distinct().Count() == a.PlanIds.Count);
                    foreach (var p in a.PlanIds) { Require(Text(p, 128)); plans.Add(p); }
                    var size = Bytes(a.Width, a.Height);
                    switch (a.Ownership)
                    {
                        case ResourceOwnership.OWNED:
                            Require(a.BorrowedKey == null && a.Sources.Count > 0 && size <= TEXTURE_BYTES);
                            facts = checked(facts + a.Sources.Count); Require(facts <= MAX_SOURCE_FACTS);
                            Require(a.Sources[0] != null); var c = Identity(a.Sources[0]); Require(contents.Add(c));
                            foreach (var v in a.Sources)
                            {
                                Require(Valid(v) && Identity(v) == c && v.Width == a.Width && v.Height == a.Height);
                                Require(canonical.Add((v.ProofScope, v.CanonicalSource)));
                            }
                            resident = checked(resident + size); owned = checked(owned + 1);
                            foreach (var p in a.PlanIds)
                            {
                                planBytes.TryGetValue(p, out var b); planCounts.TryGetValue(p, out var n);
                                planBytes[p] = checked(b + size); planCounts[p] = checked(n + 1);
                            }
                            break;
                        case ResourceOwnership.BORROWED:
                            Require(a.Sources.Count == 0 && Text(a.BorrowedKey, 4096) && borrowed.Add(a.BorrowedKey));
                            Require(a.Phase == ResourceAllocationPhase.READY); borrowedCount = checked(borrowedCount + 1);
                            break;
                        default: Require(false); break;
                    }
                }
                Require(plans.Count <= MAX_PLANS && owned <= PROCESS_ALLOCATIONS && borrowedCount <= PROCESS_ALLOCATIONS);
                Require(planBytes.Values.All(x => x <= PLAN_BYTES) && planCounts.Values.All(x => x <= PLAN_ALLOCATIONS));
                foreach (var r in state.References)
                {
                    Require(r != null); Id(r.Id);
                    Require(Kind(r.Kind) && byId.TryGetValue(r.AllocationId, out var a) && a.PlanIds.Contains(r.PlanId));
                }
                var scratch = state.Scratch;
                if (scratch != null)
                {
                    Require(scratch.Operation != null); Id(scratch.Operation.OperationId);
                    Require(byId.TryGetValue(scratch.Operation.AllocationId, out var a) && a.Ownership == ResourceOwnership.OWNED && a.PlanIds.Contains(scratch.Operation.PlanId));
                    Require(scratch.Bytes == Bytes(a.Width, a.Height) && scratch.Bytes <= TEXTURE_BYTES);
                    Require(a.Phase == (scratch.DecodeCompleted ? ResourceAllocationPhase.READY : ResourceAllocationPhase.RESERVED));
                }
                Require(state.Allocations.Where(a => a.Phase == ResourceAllocationPhase.RESERVED).All(a =>
                    scratch != null && !scratch.DecodeCompleted && scratch.Operation.AllocationId == a.Id));
                var peak = checked(resident + (scratch?.Bytes ?? 0)); Require(peak <= PROCESS_BYTES);
                return new ResourceTotals(resident, peak, owned, planBytes, planCounts);
            }
            catch (ArgumentException) { return null; }
            catch (OverflowException) { return null; }
        }
        public ResourceDecision Decide(ResourceState state, ResourceEvent e)
        {
            ResourceDecision Reject(string code) => new(state, false, code);
            if (Totals(state) == null) return Reject("INVALID_STATE");
            try
            {
                switch (e)
                {
                    case ResourceEvent.Acquire acquire:
                    {
                        if (!Text(acquire.PlanId, 128) || !Valid(acquire.Source) || !Kind(acquire.Kind)) return Reject("INVALID_SOURCE");
                        var c = Identity(acquire.Source); var owned = state.Allocations.Where(a => a.Ownership == ResourceOwnership.OWNED).ToArray();
                        if (owned.Any(a => a.Sources.Any(v => v.ProofScope == acquire.Source.ProofScope && v.CanonicalSource == acquire.Source.CanonicalSource && Identity(v) != c)))
                            return Reject("CONTRADICTORY_SOURCE");
                        var existing = owned.FirstOrDefault(a => Identity(a.Sources[0]) == c);
                        if (existing == null && state.Scratch != null) return Reject("SCRATCH_BUSY");
                        var aid = existing?.Id ?? state.NextId; var rid = existing == null ? checked(state.NextId + 1) : state.NextId;
                        var reference = new ResourceReference(rid, acquire.PlanId, aid, acquire.Kind);
                        var commands = new List<ResourceCommand> { new ResourceCommand.ReferenceGranted(reference) };
                        var updated = new ResourceAllocation(aid, ResourceOwnership.OWNED, acquire.Source.Width, acquire.Source.Height,
                            (existing?.Sources ?? Array.Empty<VerifiedResourceSource>()).Append(acquire.Source).Distinct(),
                            (existing?.PlanIds ?? Array.Empty<string>()).Append(acquire.PlanId).Distinct(), phase: existing?.Phase ?? ResourceAllocationPhase.RESERVED);
                        var scratch = state.Scratch; var next = checked(rid + 1);
                        if (existing == null)
                        {
                            var op = new ResourceDecodeOperation(next, aid, acquire.PlanId); next = checked(next + 1);
                            scratch = new ResourceScratch(op, Bytes(updated.Width, updated.Height));
                            commands.Add(new ResourceCommand.Decode(op, acquire.Source));
                        }
                        return Admit(state, new ResourceState(state.Allocations.Where(a => a.Id != aid).Append(updated),
                            state.References.Append(reference), scratch, next), commands.ToArray());
                    }
                    case ResourceEvent.AcquireBorrowed acquire:
                    {
                        if (!Text(acquire.PlanId, 128) || !Text(acquire.BorrowedKey, 4096) || !Kind(acquire.Kind)) return Reject("INVALID_BORROWED");
                        Bytes(acquire.Width, acquire.Height);
                        var existing = state.Allocations.FirstOrDefault(a => a.Ownership == ResourceOwnership.BORROWED && a.BorrowedKey == acquire.BorrowedKey);
                        if (existing != null && (existing.Width != acquire.Width || existing.Height != acquire.Height)) return Reject("CONTRADICTORY_BORROWED");
                        var aid = existing?.Id ?? state.NextId; var rid = existing == null ? checked(state.NextId + 1) : state.NextId;
                        var reference = new ResourceReference(rid, acquire.PlanId, aid, acquire.Kind);
                        var allocation = new ResourceAllocation(aid, ResourceOwnership.BORROWED, acquire.Width, acquire.Height,
                            planIds: (existing?.PlanIds ?? Array.Empty<string>()).Append(acquire.PlanId).Distinct(),
                            borrowedKey: acquire.BorrowedKey, phase: ResourceAllocationPhase.READY);
                        return Admit(state, new ResourceState(state.Allocations.Where(a => a.Id != aid).Append(allocation),
                            state.References.Append(reference), state.Scratch, checked(rid + 1)), new ResourceCommand.ReferenceGranted(reference));
                    }
                    case ResourceEvent.ReleaseReference release:
                        if (release.Reference == null || !state.References.Contains(release.Reference)) return Reject("STALE_REFERENCE");
                        return Admit(state, new ResourceState(state.Allocations, state.References.Where(r => r != release.Reference), state.Scratch, state.NextId));
                    case ResourceEvent.DecodeCompleted completed:
                    {
                        var scratch = state.Scratch;
                        if (scratch == null || scratch.Operation != completed.Operation || scratch.DecodeCompleted) return Reject("STALE_DECODE");
                        var allocations = state.Allocations.Select(a => a.Id != completed.Operation.AllocationId ? a :
                            new ResourceAllocation(a.Id, a.Ownership, a.Width, a.Height, a.Sources, a.PlanIds, a.BorrowedKey, ResourceAllocationPhase.READY));
                        return Admit(state, new ResourceState(allocations, state.References, scratch with { DecodeCompleted = true }, state.NextId),
                            new ResourceCommand.ReleaseUploadScratch(completed.Operation));
                    }
                    case ResourceEvent.ScratchReleased released:
                    {
                        var scratch = state.Scratch;
                        if (scratch == null || scratch.Operation != released.Operation || !scratch.DecodeCompleted) return Reject("STALE_SCRATCH");
                        return Admit(state, new ResourceState(state.Allocations, state.References, null, state.NextId));
                    }
                    default: return Reject("INVALID_EVENT");
                }
            }
            catch (ArgumentException) { return Reject("INVALID_EVENT"); }
            catch (OverflowException) { return Reject("OVERFLOW"); }
        }
        private ResourceDecision Admit(ResourceState prior, ResourceState next, params ResourceCommand[] commands) =>
            Totals(next) == null ? new ResourceDecision(prior, false, "LIMIT_OR_INCONSISTENT") : new ResourceDecision(next, true, "ACCEPTED", commands);
    }
}
